.github/workflows/codeql.yml

```diff
@@ -0,0 +1,82 @@
+# For most projects, this workflow file will not need changing; you simply need
+# to commit it to your repository.
+#
+# You may wish to alter this file to override the set of languages analyzed,
+# or to provide custom queries or build logic.
+#
+# ******** NOTE ********
+# We have attempted to detect the languages in your repository. Please check
+# the `language` matrix defined below to confirm you have the correct set of
+# supported CodeQL languages.
+#
+name: "CodeQL"
+
+on:
+  push:
+    branches: [ "master" ]
+  pull_request:
+    # The branches below must be a subset of the branches above
+    branches: [ "master" ]
+  schedule:
+    - cron: '45 21 * * 6'
+
+jobs:
+  analyze:
+    name: Analyze
+    # Runner size impacts CodeQL analysis time. To learn more, please see:
+    #   - https://gh.io/recommended-hardware-resources-for-running-codeql
+    #   - https://gh.io/supported-runners-and-hardware-resources
+    #   - https://gh.io/using-larger-runners
+    # Consider using larger runners for possible analysis time improvements.
+    runs-on: ${{ (matrix.language == 'swift' && 'macos-latest') || 'ubuntu-latest' }}
+    timeout-minutes: ${{ (matrix.language == 'swift' && 120) || 360 }}
+    permissions:
+      actions: read
+      contents: read
+      security-events: write
+
+    strategy:
+      fail-fast: false
+      matrix:
+        language: [ 'go' ]
+        # CodeQL supports [ 'c-cpp', 'csharp', 'go', 'java-kotlin', 'javascript-typescript', 'python', 'ruby', 'swift' ]
+        # Use only 'java-kotlin' to analyze code written in Java, Kotlin or both
+        # Use only 'javascript-typescript' to analyze code written in JavaScript, TypeScript or both
+        # Learn more about CodeQL language support at https://aka.ms/codeql-docs/language-support
+
+    steps:
+    - name: Checkout repository
+      uses: actions/checkout@v3
+
+    # Initializes the CodeQL tools for scanning.
+    - name: Initialize CodeQL
+      uses: github/codeql-action/init@v2
+      with:
+        languages: ${{ matrix.language }}
+        # If you wish to specify custom queries, you can do so here or in a config file.
+        # By default, queries listed here will override any specified in a config file.
+        # Prefix the list here with "+" to use these queries and those in the config file.
+
+        # For more details on CodeQL's query packs, refer to: https://docs.github.com/en/code-security/code-scanning/automatically-scanning-your-code-for-vulnerabilities-and-errors/configuring-code-scanning#using-queries-in-ql-packs
+        # queries: security-extended,security-and-quality
+
+
+    # Autobuild attempts to build any compiled languages (C/C++, C#, Go, Java, or Swift).
+    # If this step fails, then you should remove it and run the build manually (see below)
+    - name: Autobuild
+      uses: github/codeql-action/autobuild@v2
```

Comment:
This part feels like a lot of magic; I suspect it will just do a `go build .` (or along those lines), which won't match how we build our actual binaries (we may be passing specific build options).

Wondering if we can integrate this with a build-stage, and provide it access to the produced artefacts instead.

I see the lines below this outline "here's what to do if you don't want to use the auto build"

I suspect `buildx` and `buildkit` will be in a similar situation, and @crazy-max has a lot more GHA-Fu than I have; perhaps he has suggestions (and we can take a similar approach in the buildx repository - not having to reinvent the wheel)

Reply:
Actually autobuild tries to detect build scripts and common build systems, so from the logs it look like it's using the Makefile (https://github.com/docker/cli/actions/runs/6641001271/job/18042547995?pr=4624#step:4:110)

As for giving it access to the produced build artefacts, that wouldn't do the trick as it does some tracing and instrumentation during compilation to do the extraction (https://docs.github.com/en/code-security/codeql-cli/codeql-cli-manual/database-create#-c---commandcommand). Same reason [caching is an issue](https://github.com/orgs/community/discussions/28654#discussioncomment-3708403).

Reply:
Hm, right. One thing I notice at least that in that case, it builds on the host (not using the Dockerfile), and will build with a different version of Go;

```
2023/10/25 13:19:30 Autobuilder was built with go1.21.1, environment has go1.20.10
```

So we may need a `go-setup` action added to have the same version of Go as other workflows at least.

Do we know if the CodeQL analysis can be injected into a container somehow, so that it runs in the same environment as we'd otherwise be running in? 🤔

Reply:
https://docs.github.com/en/code-security/code-scanning/creating-an-advanced-setup-for-code-scanning/running-codeql-code-scanning-in-a-container should be doable

Reply:
Hm, actually, for this to work like in the example
```
    # Specify the container in which actions will run
    container:
      image: codeql-container:f0f91db
```
it would need to be an image from Hub or another configured registry, can't just give it a Dockerfile